[FE] add tldraw

pigpgw opened the pull request:
## 주요 변경사항
- 그림판 기능을 추가했습니다.
- 지속적인 리펙토링을 하고있습니다.
## 리뷰어에게...

## 관련 이슈

- closes #44

pigpgw committed:
feat: add DrawBoard function

pigpgw committed:
remove: remove unnecessary code

pigpgw committed:
remove: remove unnecessary code and refactor button click event name

pigpgw committed:
refactor: refactor DrawBoard components

pigpgw committed:
refactor: file components

loopy-lim commented on src/components/Draw/DrawBoard.tsx:
```diff
@@ -0,0 +1,52 @@
+const NameEditor = track(() => {
+  const editor = useEditor();
+
+  const { color, name } = editor.user.getUserPreferences();
```
아무리 그래도 name은 가지고 와서 넣어줍시다

pigpgw replied:
크흠...

loopy-lim commented on src/components/Draw/DrawBoard.tsx:
```diff
@@ -0,0 +1,52 @@
+export const DrawBoard = ({ roomId }: { roomId: string }) => {
+  const store = useTldrawStore({
+    roomId,
+    hostUrl: "wss://demos.yjs.dev/ws",
```
금방 하겠습니다..허허

pigpgw replied:
언제나 감사합니다.

loopy-lim commented on src/components/Draw/DrawBoard.tsx:
```diff
@@ -0,0 +1,52 @@
+  );
+});
+
+export const DrawBoard = ({ roomId }: { roomId: string }) => {
```
interface?

pigpgw replied:
훠훠

loopy-lim commented on src/components/Draw/DrawBoard.tsx:
```diff
@@ -0,0 +1,52 @@
+
+  return (
+    <div className="h-full w-full">
+      <div className="tldraw__editor h-full w-full">
```
w-full? div의 기본 속성은 무엇일까요?

pigpgw replied:
display block (번지점프 하는중)

loopy-lim commented on src/components/File/PrSelectedFileVier/PrFileNameViewer.tsx:
```diff
@@ -0,0 +1,7 @@
+export const PrFileNameViewer = ({ fileName }: { fileName: string }) => {
```
interface....?

pigpgw replied:
어이쿠 고양이가 쳤나봐요...

loopy-lim commented on src/components/File/SelectedFileViewer.tsx:
```diff
@@ -128,21 +128,23 @@ const InitFile = ({
+          {commitFileList.map((file, index) => {
             return (
-              <TableRow>
+              <TableRow key={index} className="text-4xl">
```
이게 옳게된 세상이다

pigpgw replied:
저를 올바른 길로 인도해주셔서 감사합니다... shadcn은 key warning이 안뜨는 것 같다. 아마도...

loopy-lim commented on src/components/GNB/LeftGNB.tsx:
```diff
@@ -1,7 +1,11 @@
 import { Folder, MessageSquare, NotepadText, SquarePen } from "lucide-react";
 
-const LeftGNB = () => {
+interface LeftGNBProps {
+  toggleDrawBoard(): void;
+}
```
이거와 관련해서는 toggleDrawBoard는 정말 좋은 겁니다. 하지만 저희 스펙상 note라는 것도 있기 때문에, 어떻게 하면 좋을지 고민을 많이 하신 것 같아요.

그래서 toggleDrawBaord가 아니라` navigateFullLeftGNB("value")`가 어떤가...?라는 생각입니다.

pigpgw replied:
감사합니다. 항상 네이밍에 고민이 많아서 여러번 고민했는데 아직도 어렵네요 좋은 의견 감사합니다.

loopy-lim commented on src/hooks/useTldrawStore.tsx:
```diff
@@ -0,0 +1,327 @@
+  version: number;
+  shapeUtils: TLAnyShapeUtilConstructor[];
+}>) {
+  const [store] = useState(() => {
+    const store = createTLStore({
+      shapeUtils: [...defaultShapeUtils, ...shapeUtils],
+    });
+
+    return store;
+  });
```
좋은 사용방법이에요. 왜일까요? 한번 찾아보시고 댓글 달아주시면 좋을것 같습니다.

pigpgw replied:
lazy loading을 알고있었는데 state에서도 사용하는 것인지는 처음 알았습니다. 해당 컴포넌트가 너무 커서 조금 지연되서 보여도 되는 컴포넌트나 이미지같은 경우 지연로딩을 사용한다 정도만 알고있었는데 이렇게 state에서도 사용이 가능한지는 몰랐습니다.  아래는  제가 찾아본 지연 로딩 개념입니다.

> useState에 직접적인 값 대신에 함수를 넘기는 것을 게으른 초기화(lazy)라고 한다. react 공식 문서에서는, 이러한 게으른 초기화를 초기 값이 복잡한 연산을 포함할 때 사용하라고 되어 있다. 게으른 초기화 함수는 오직 state가 처음 만들어 질 때만 실행된다. 이 후에 다시 리렌더링이 된다면, 이 함수의 실행은 무시된다.

처음에는 이 당연한 말이 조금 이해가 안됬었습니다. "그게 안되도 괜찮나?"라는  말도 안되는 생각을 잠시 했었습니다. 리렌더링이란 상태 변화에따른 상태값 업데이트인데 피곤해서 그런지 리렌더링마다 계산 안되는거 아닌가? 아니 이거 계산되면 상태값 업데이트는 어떻게하는거야? 라는 생각을 했었는데 찾아보니  "컴포넌트가 렌더링될 때마다 초기화에 등록된 코드가 실행되는데 결과값은 사용되지 않지만 자바스크립트 특성상 함수 호출 자체는 일어나서 실행은되버린다" 라는걸 알았습니다. 근데 만약 해당 코드가 계산이 가볍다면 상관이 없겠지만 무거운 코드라면 매 리렌더링마다 계산이 일어난다면 사용자 경험을 저하시키는(성능 낭비) 문제를 유발할 수 있다는 것을 알게되었습니다. 다만 주의할 점은 불필요한 경우 사용하면 오히려 코드의 가독성을 저하시키고 코드가 깔끔하지 못하다는 단점이 있다는걸 알게되었습니다. 감사합니다.

loopy-lim replied:
@pigpgw 
이제야 읽게 되었네요. 좋은 방법이라는 것이라는 것을 알다니 좋습니다.
특히 렌더링과 상태에따라서 동작하는 것이 다를 때가 있기 때문에 이 사용방법에 대해서 알고 있다면 추가적인 선택지가 될 수 있겠다라는 생각을 하게 됩니다.

loopy-lim commented on src/hooks/useTldrawStore.tsx:
```diff
@@ -0,0 +1,327 @@
+  });
+
+  const { yDoc, yStore, meta, room } = useMemo(() => {
+    const yDoc = new Y.Doc({ gc: true });
```
gc가 필요할까요? 그리고 yDoc저희 쓰고 있지 않나요?

pigpgw replied:
- 문서 변경 이력에서 불필요한 데이터(오래된 변경 이력)을 정리(자동 삭제)하는 기능이라고 알고있는데  불필요한 이유가 궁금합니다.!
- yDoc은 쓰고있지만 다른 기능이라고 생각해서 합치지는 않았습니다. 물론 한가지 yDoc만 쓰는것은 가능한데 기능이 다르기에 관리하는 측면에서도 분리하는게 좋다고 생각하는데 하나로 써야한다고 생각하신 이유가 궁금합니다.!

loopy-lim replied:
@pigpgw 
지금에서야 말하자면 동의가 될 것이라고 생각될 것입니다.
Ydoc이 분리되지 않는 이유는 아시다 싶이 Source가 분리되면 관리되기 어렵기 때문입니다.
그리고 사실 gc는 지금은 필요없다고 생각했어요. (그냥 폴리싱때나 느려질때 추가하는 것이 좋지 않을까? 라는 생각이었어요)

pigpgw replied:
생각해보니 그렇네요 관리가 어렵다 그걸 저도 에저 느껴서 공감이 됬습니다. 감사합니다. gc 확인

loopy-lim commented on src/hooks/useTldrawStore.tsx:
```diff
@@ -0,0 +1,327 @@
+
+  const { yDoc, yStore, meta, room } = useMemo(() => {
+    const yDoc = new Y.Doc({ gc: true });
+    const yArr = yDoc.getArray<{ key: string; val: TLRecord }>(`tl_${roomId}`);
```
마찬가지로 yArr도 저희가 쓸 수 있을꺼 같은데...

pigpgw replied:
저희가 쓸 수 있다리... 크흠... 혹시 조금만 더 자세히 설명해 주실 수 있으신...

loopy-lim replied:
따로 커스텀하게 사용할 수 있을 것 같다라는 생각이었는데, 이미 지금 사용하고 계시군요

pigpgw replied:
요시

loopy-lim commented on src/hooks/useTldrawStore.tsx:
```diff
@@ -0,0 +1,327 @@
+        const toPut: TLInstancePresence[] = [];
+
+        // Connect records to put / remove
+        for (const clientId of update.added) {
+          const state = states.get(clientId);
+          if (state?.presence && state.presence.id !== presenceId) {
+            toPut.push(state.presence);
+          }
```
코드가 좋네요 맛있어요

loopy-lim commented on src/hooks/useTldrawStore.tsx:
그래도 주석으로 어디서 가지고 왔는지 알려주면 좋을 것 같아요

pigpgw replied:
넵 알겠습니다. 추가하도록 하겠습니다.

loopy-lim commented on src/pages/conversation/ConversationPage.tsx:
```diff
@@ -49,6 +52,8 @@ const ConversationPage = () => {
   const bindingRef = useRef<MonacoBinding | null>(null);
   const fileMetadata = ydoc.getArray<PrChangedFileInfo>("fileMetadata");
 
+  const [drawBoard, setDrawBoard] = useState(false);
```
store라는 좋은 것도 있습니다

pigpgw replied:
감사합니다!

loopy-lim commented on src/pages/conversation/ConversationPage.tsx:
```diff
@@ -170,23 +175,27 @@ const ConversationPage = () => {
   }, [ydoc, editor, selectedCommitFile]);
 
-  const handleEditorMount = (editorInstance: editor.IStandaloneCodeEditor) => {
+  const onEditorMount = (editorInstance: editor.IStandaloneCodeEditor) => {
```
감사합니다

pigpgw replied:
제가 더 감사합니다.

loopy-lim commented on src/pages/conversation/ConversationPage.tsx:
```diff
@@ -170,23 +175,27 @@ const ConversationPage = () => {
     setEditor(modifiedEditor);
   };
 
+  const toggleDrawBoard = () => {
+    setDrawBoard((prev) => !prev);
+  };
+
```
toggle을 쓰고 싶다면 다음과 같은 코드로 쓸 수 있어요~

```suggestion
const [drawBoard, toggleDrawBoard] = useReducer(false, (state) => !state);
```

loopy-lim replied:
저렇게쓰면 사이드이펙트가 줄어들겟죠?
그리고 코드가 짧고 오히려 익숙해진다면 알기가 더 편해질꺼에요

pigpgw replied:
감사합니다. 이번기회에 reducer에 대해서 공부해보겠습니다.!

loopy-lim reviewed (approved):
좋아유~
몇개는 바꾸셔도 좋지만, 머... 이름만은 좀 바꾸어 주고 merge하셔도 될 것 같습니다.

pigpgw committed:
refactor: refactor DrawBoard user name

pigpgw committed:
remove: remove unnecessary code

pigpgw committed:
refactor:  add PrFilePathViewerProps , PrFileNameViewerProps interface

pigpgw committed:
docs: Add source attribution for tldraw implementation

body:
- Add source attribution comments for code adapted from tldraw-yjs-example
- Original source: https://github.com/tldraw/tldraw-yjs-example
- This implementation provides real-time collaborative drawing features using tldraw and Yjs
- Added MIT license reference and documentation for future maintainability

pigpgw reviewed (commented):
새해 복 많이받으세요. 소중한 리뷰 감사합니다.
